Add new Camera3D type: Oblique

qaptoR opened the pull request:
Closes: https://github.com/godotengine/godot-proposals/issues/501
Partial solution to this proposal: https://github.com/godotengine/godot-proposals/issues/2713

This PR should not necessarily be a reason to close the proposal 2713, it is merely a straightforward implementation of oblique near plane clipping to implement a portal--or other similar--mechanic.

This PR is essentially an update to my previous PR: https://github.com/godotengine/godot/pull/64876
The main difference is that shadows were fixed.

This other PR attempts to completely solve proposal 2713 mentioned above, but apparently is having issues with shadows. Considering that its commits try to implement custom projections by adding a new camera type like with my PRs, there is a good chance the way I fixed shadows could also work there too: https://github.com/godotengine/godot/pull/84454

The solution was that along with the main Oblique camera projection a separate, normal Perspective camera projection needed to be included with the `CameraData` used in the `RenderSceneCull::_render_scene()` that was used explicitly for lighting and shadows.

However, regarding the potential question of whether one PR should be included over the other as a solution, clearly the more general solution can see a wider range of uses. However, I believe this implementation of an Oblique Near Plane camera still has merit based on the majority use case of creating a portal mechanic, etc where oblique near clipping is necessary.

This is because the interface for the Oblique camera included in this PR has been simplified so that users are only required to provide at a minimum the transform of the plane they wish to use as the near clipping plane of the camera using `set_oblique_plane_from_transform()`.

Additionally, because all of the required math (especially the matrix math) is happening in C++, the effect is as efficient as possible, without requiring steps to first `camera.get_projection`, then doing matrix manipulation, then `camera.set_projection` all in GDScript.

Alternatively, because the Oblique Near Plane matrix manipulations "... can be applied to any projection matrix, ..." (https://terathon.com/lengyel/Lengyel-Oblique.pdf, pg. 2) it's possible that all camera types (including the other proposed custom projection camera) could benefit from a general `is_oblique_near_plane` boolean toggle instead of this dedicated oblique camera which only implements the algorithm for a Perspective projection.

Video 1: Current state of personal portal game mechanic
https://youtu.be/A_-2nTyKVkU
Notice that the portal viewports are not blocked by the surface the portal is on (the issue with using a standard perspective camera), and that shadows and lighting appear the same when looking through the portal (as evidenced by walking through the portal where a seamless transition occurs).

Video 2: Previous video showcasing dynamically the difference oblique near plane clipping makes
https://youtu.be/PipVd6wRnMk?si=lCThReaKyO-smBaP
Note that this older video did not properly handle shadows, and the viewport texture color and environment lighting were all wrong giving it that washed out look (both fixed in Video 1).

Image 1: Portal effect using Perspective camera
![image](https://github.com/godotengine/godot/assets/40372043/9d405b1a-6dab-4ab0-a72f-6a1d7765d8ca)

Image 2: Portal effect using Oblique camera
![image](https://github.com/godotengine/godot/assets/40372043/c7c52a6b-17be-4703-bb29-7bb0e4e5f843)

Image 3: Oblique camera editor interface
![image](https://github.com/godotengine/godot/assets/40372043/7e4a1896-324b-4bd0-96d7-44c463594f3f)

Commit:
Implementation of new Camera3D type: Oblique
allows the specification of a world plane for near clipping;
applies matrix manipulation to generate oblique near plane.

AThousandShips commented on core/math/projection.h:
```diff
@@ -31,6 +31,7 @@
 #ifndef PROJECTION_H
 #define PROJECTION_H
 
+#include "core/math/quaternion.h"
```
Should be forward declared instead here

AThousandShips commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -71,6 +72,20 @@ void RendererSceneCull::camera_set_perspective(RID p_camera, float p_fovy_degree
 
+void RendererSceneCull::camera_set_oblique(RID p_camera, float p_fovy_degrees, const Vector3 &p_ob_normal, const Vector3 &p_ob_position, float p_ob_offset, float p_z_near, float p_z_far) {
+	Camera *camera = camera_owner.get_or_null(p_camera);
+	ERR_FAIL_COND(!camera);
```
```suggestion
	ERR_FAIL_NULL(camera);
```

AThousandShips commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -31,6 +31,7 @@
 #include "renderer_scene_cull.h"
 
 #include "core/config/project_settings.h"
+#include "core/math/quaternion.h"
```
```suggestion
```
Already included in the header

AThousandShips commented on scene/3d/camera_3d.h:
```diff
@@ -31,6 +31,7 @@
 #ifndef CAMERA_3D_H
 #define CAMERA_3D_H
 
+#include "core/math/quaternion.h"
```
```suggestion
```
Already included

Calinou reviewed (commented):
Tested locally, I'm having trouble setting up a planar water reflection with this PR on Linux + NVIDIA.

I get frequent engine freezes (with `VK_DEVICE_LOST` error and the entire PC appearing to freeze) if invalid transforms are entered in the Oblique camera projection mode, so it's hard to test. I had the same problems when testing https://github.com/godotengine/godot/pull/84454.

PS: I wonder if setting projection from a transform should be offered in the editor somehow, so you can set up portals and planar reflections without any script. The RemoteTransform3D node makes it possible to adjust the camera position without a script, but I still need to assign a script to call `Camera3D.set_oblique_plane_from_transform()`. Ideally, it should be possible to set this up without any script to make this feature more accessible to artists and level designers.

**Testing project:** [test_pr_89140.zip](https://github.com/godotengine/godot/files/14484253/test_pr_89140.zip)

qaptoR commented:
@Calinou 
There are two camera member variables: oblique_normal, and oblique_position, which are exposed in the editor for this purpose. 
Use the transform origin for position and transform forward vector for the normal.

Calinou commented:
@qaptoR Can you upload a testing project? I haven't been able to figure it out so far.

qaptoR commented:
@Calinou 
I just got home from classes. 
 - I will post a minimum testing project asap. 
 - I will also test the project you posted on my system to see if everything works: 
   - I'm building/testing on Windows 10, Intel 10900K, Nvidia 3090.
 - then work on the code revisions that have been posted so far.

qaptoR commented:
@Calinou 
Still going to work on a minimum test project, but in the interim I figured out what was wrong with your test scene.

The path to the `Camera3D` in `water.gd` on line 12 should be:
`$Water/SubViewport/Camera3D.set_oblique_plane_from_transform(global_transform)`

I was unable to open the test scene until this was corrected because it is an `@tool` script.
From there the oblique camera works correctly for me.

fire commented:
@lyuma You were looking into this.

qaptoR commented:
@Calinou
Here is a minimum portal test project.
Some things to note about the scene I set up
- the BackingMesh's are invisible to the portal cameras purposefully because they're mostly there to keep tabs on where the portals are as you move around
- when you look through the portal at the other portal, the visual distortion that looks like you're painting on the portal with what else is in view is normal because we're not handling recursion
- there are visual indicators behind each portal so that when you switch the cameras to perspective mode it's obvious what the oblique camera is culling

[89104_min_test.zip](https://github.com/godotengine/godot/files/14487622/89104_min_test.zip)

qaptoR commented on core/math/projection.cpp:
```diff
@@ -313,6 +314,59 @@ void Projection::set_perspective(real_t p_fovy_degrees, real_t p_aspect, real_t
+	Quaternion q;
+
+	/* clang-format off */
+	q.x = ((
+		p_oblique_plane.x < 0 ? -1 :
+		p_oblique_plane.x > 0 ? +1 : 0
+		) + columns[2][0]
+	) / columns[0][0];
+
+	q.y = ((
+		p_oblique_plane.y < 0 ? -1 :
+		p_oblique_plane.y > 0 ? +1 : 0
+		) + columns[2][1]
+	) / columns[1][1];
```
Updated code to:
```
q.x = (SIGN(p_oblique_plane.x) + columns[2][0]) / columns[0][0];
q.y = (SIGN(p_oblique_plane.y) + columns[2][1]) / columns[1][1];
```

from what I recall, when I originally wrote the first version of this camera almost 3 years ago the `SIGN` macro didn't exist. Either that or at the time I did not search thoroughly enough. Given how quickly I found it now however it could be that either it was the first case, or I'm just more familiar with the engine codebase that finding things is easier.

lyuma commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -2591,7 +2617,23 @@ void RendererSceneCull::render_camera(const Ref<RenderSceneBuffers> &p_render_bu
 						camera->vaspect);
+			} break;
+			case Camera::OBLIQUE: {
+				Quaternion oblique_plane = get_camera_oblique_quaternion(p_camera);
```
If this is a plane, not a quaternion, then the function should be named `get_camera_oblique_plane` and the return type should be `Plane`, not `Quaternion`.

If Godot's Plane type is somehow not suitable for this case, then it should use Vector4 rather than Quaternion.

qaptoR replied:
The drawback to using a plane is that must be converted to quaternion before it can be used in the step `p_oblique_plane * (2.0F / p_oblique_plane.dot(q))` because the Plane type does not support mathematical operations.

Quaternion was used because at the time (almost 3 years ago) that this camera was originally implemented the Vector4 type did not exist.

lyuma commented on editor/plugins/gizmos/camera_3d_gizmo_plugin.cpp:
```diff
@@ -212,6 +212,28 @@ void Camera3DGizmoPlugin::redraw(EditorNode3DGizmo *p_gizmo) {
 			ADD_TRIANGLE(tup, side + up, nside + up);
 		} break;
 
+		case Camera3D::PROJECTION_OBLIQUE: {
+			// Mirrors the PROJECTION_PERSPECTIVE because they are fundamentally the same.
+			float fov = camera->get_fov() / 2.0;
+
```
If this is the same code as `PROJECTION_PERSPECTIVE` then the code should be shared using multiple case statements.
```
case Camera3D::PROJECTION_PERSPECTIVE:
case Camera3D::PROJECTION_OBLIQUE:
{
```

qaptoR replied:
this will be updated in the next push

lyuma commented on doc/classes/Camera3D.xml:
```diff
@@ -202,6 +221,15 @@
 		<member name="near" type="float" setter="set_near" getter="get_near" default="0.05">
 			The distance to the near culling boundary for this camera relative to its local Z axis. Lower values allow the camera to see objects more up close to its origin, at the cost of lower precision across the [i]entire[/i] range. Values lower than the default can lead to increased Z-fighting.
 		</member>
+		<member name="oblique_normal" type="Vector3" setter="set_oblique_normal" getter="get_oblique_normal" default="Vector3(0, 1, 0)">
+			The desired normal vector of the world space plane used by an oblique camera as an oblique near clipping plane.
+		</member>
+		<member name="oblique_offset" type="float" setter="set_oblique_offset" getter="get_oblique_offset" default="0.0">
+			The offset value for the oblique plane along its forward vector.
+		</member>
```
I'm not convinced about this using a separate projection mode from perspective. Most of the code seems to mimic perspective, just with some extra options. Perhaps we can condition this on whether those extra options are in use (for example, if `oblique_offset == Vector3.ZERO` and `oblique_normal == Vector3(0,1,0)` then it is a standard perspective. (is that default value of (0,1,0) correct?)

qaptoR replied:
I've mentioned this in the top PR comment, which is that Lengyel's algorithm can technically be applied to any projection matrix, and it will remain the same except for the changes to the near clipping plane. My suggestions is to add a boolean flag to the camera which appends the matrix manipulation algorithm to the end of all the camera types. Rather than duplicating a single camera type.

This would also further justify the added second `shadow_projection` included in the `CameraData` which is used to maintain the shadows and lighting, as it is currently overhead which is only necessitated by the Oblique camera.
If all camera types had the option of Oblique near plane projections then it would be further justified.

lyuma replied:
> My suggestions is to add a boolean flag to the camera which appends the matrix manipulation algorithm to the end of all the camera types. Rather than duplicating a single camera type.

it sounds reasonable to me. we just have to be careful not to add overhead in the common case (not using oblique)

lyuma commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -3155,7 +3201,7 @@ void RendererSceneCull::_render_scene(const RendererSceneRender::CameraData *p_c
 		cull_data.render_reflection_probe = render_reflection_probe;
 		cull_data.occlusion_buffer = RendererSceneOcclusionCull::get_singleton()->buffer_get_ptr(p_viewport);
-		cull_data.camera_matrix = &p_camera_data->main_projection;
+		cull_data.camera_matrix = &p_camera_data->shadow_projection;
```
Why is it correct for culling to use the plain perspective projection?

If the intent is to use this for planar reflections, wouldn't a frustum camera projection be more appropriate for culling purposes?

qaptoR replied:
The oblique camera is only implemented as essentially a 'trick' version of the Perspective camera. This is because it was developed with games that use that camera in mind.

As mentioned in an above comment, the `shadow_projection` as overhead is only in service to this one camera, for every other type it will be a duplicate of the `main_projection`, which is why I call it overhead.

I don't know enough about the rendering pipeline or lighting system to understand why using the Oblique projection breaks the shadows (a problem with my first PR for this feature), but my guess is that for every *_Projection, any attempt to use the Oblique_*_Projection variation will require the shadows and lighting to use the original *_Projection.

This is why I feel that adding the boolean toggle for all cameras would justify the overhead.

lyuma replied:
yeah I get the idea of adding a boolean to enable the overhead. I can't entirely say I understand if it's correct for light/shadow to use a separate projection matrix (might it also add complexity to the fragment shader?)

the comment I am trying to make here is that culling could be based on the frustum. I show an example here of how using the frustum projection for culling can be helpful: https://github.com/godotengine/godot/pull/85529#issuecomment-1915745899
it would be good to double check if occlusion culling works correctly since the origin point in a classical perspective projection might be outside the room.... while using a frustum might make this check more correct.

Flarkk replied:
Culling currently assumes the near and far planes are parallel. Passing the non-oblique projection probably avoids some troubles there. Now, the consequence is that objects between the oblique and the non-oblique planes won't be culled on the CPU (they will be on the GPU anyway). This is probably an acceptable trade-off though.

lyuma commented on core/math/projection.cpp:
```diff
@@ -313,6 +314,48 @@ void Projection::set_perspective(real_t p_fovy_degrees, real_t p_aspect, real_t
 }
 
+void Projection::set_oblique(real_t p_fovy_degrees, real_t p_aspect, real_t p_z_near, real_t p_z_far, Quaternion p_oblique_plane, bool p_flip_fov) {
+	if (p_flip_fov) {
+		p_fovy_degrees = get_fovy(p_fovy_degrees, 1.0 / p_aspect);
+	}
```
I don't understand what p_flip_fov is for

qaptoR replied:
this isn't part of my PR, This is just part of the Projection::set_perspective code. Almost everything about the Oblique camera is a duplication of the Perspective camera because it was the most straightforward way to implement oblique near plane clipping for my project when I originally started. This is why I've made the suggestion to switch to a more general approach using a boolean toggle to enable oblique near plane clipping for any camera projection type.

lyuma commented on core/math/projection.cpp:
```diff
@@ -313,6 +314,48 @@ void Projection::set_perspective(real_t p_fovy_degrees, real_t p_aspect, real_t
+	if (p_flip_fov) {
+		p_fovy_degrees = get_fovy(p_fovy_degrees, 1.0 / p_aspect);
+	}
+
+	real_t sine, cotangent, deltaZ;
+	real_t radians = Math::deg_to_rad(p_fovy_degrees / 2.0);
+
+	deltaZ = p_z_far - p_z_near;
+	sine = Math::sin(radians);
+
+	if ((deltaZ == 0) || (sine == 0) || (p_aspect == 0)) {
+		return;
+	}
+	cotangent = Math::cos(radians) / sine;
+
+	set_identity();
+
+	columns[0][0] = cotangent / p_aspect;
+	columns[1][1] = cotangent;
+	columns[2][2] = -(p_z_far + p_z_near) / deltaZ;
+	columns[2][3] = -1;
+	columns[3][2] = -2 * p_z_near * p_z_far / deltaZ;
+	columns[3][3] = 0;
```
Duplicate code of `set_perspective` (non-stereo overload)

Note also that for XR usage, it might be good to consider the possibility of supporting this in stereo in the future..

lyuma commented on core/math/projection.cpp:
```diff
@@ -313,6 +314,48 @@ void Projection::set_perspective(real_t p_fovy_degrees, real_t p_aspect, real_t
+	q.w = (1.0F + columns[2][2]) / columns[3][2];
+
+	Quaternion c = p_oblique_plane * (2.0F / p_oblique_plane.dot(q));
+	columns[0][2] = c.x;
+	columns[1][2] = c.y;
+	columns[2][2] = c.z + 1.0F;
+	columns[3][2] = c.w;
```
I would avoid taking a shortcut here, in case we do want to use this in XR. The math is the same (you'll be subtracting 0 for x, y and w and subtracting 1 for z, just like in your code) but this way it can be more easily refactored for other projection types in the future.
```suggestion
	columns[0][2] = c.x - columns[0][3];
	columns[1][2] = c.y - columns[1][3];
	columns[2][2] = c.z - columns[2][3];
	columns[3][2] = c.w - columns[3][3];
```

qaptoR replied:
I wasn't taking any shortcuts. These are the exact steps laid out in Eric Lengyel's algorithm (https://terathon.com/code/oblique.html).

I don't know enough linear algebra to determine if your suggestion will observe the same results, but as I quoted in the top PR comment, Eric's paper suggests that this algorithm is correct for any projection matrix, which I assume means that it can be used confidently unchanged.

lyuma replied:
The generalization is explained tersely here: https://aras-p.info/texts/obliqueortho.html
from what it sounds like, this correction enables the math to work for any projection matrix, not just perspective.

lyuma commented on core/math/projection.h:
```diff
@@ -34,6 +34,8 @@
 #include "core/math/vector3.h"
 #include "core/math/vector4.h"
 
+struct Quaternion;
+
```
don't forward declare Quaternion.

As I mention elsewhere, I would use Plane for this since it literally is used to refer to a Plane, but if Plane is somehow not working, use a Vector4. Don't use Quaternion.

lyuma reviewed (commented):
Interesting approach. At least in non-XR cases, I could see this being a viable alternative to my PR...
* #85529

I'm a bit unclear in how to apply this design to XR, since I'll need to copy the projection matrix of the HMD and then make it oblique.  See [my reference XR mirror implementation](https://github.com/V-Sekai/avatar_vr_demo/blob/master/addons/V-Sekai.xr-mirror/mirror.gd)

lyuma commented:
another important feature in addition to portals is Planar reflections, which require the ability to flip the final projection matrix (UV space), since this is necessary to preserve chirality of triangles (backface culling) in a mirrored setting.

for example, in my code, I do this by adding a flip in the projection here:
```
p_cam.set("override_projection",  Projection(Transform3D.FLIP_X) * proj *  Projection(Transform3D.FLIP_X))
```
and then perform the flip again in the shader to undo this:
```
		uv = vec2(1.0 - SCREEN_UV.x, SCREEN_UV.y);
```

Without this flipping, it is quite complicated to implement mirrors: there will need to be some way to communicate that culling is inverted (In [Unity users do this with `GL.invertCulling = true`](https://docs.unity3d.com/ScriptReference/GL-invertCulling.html) but inverting UV space is IMHO far simpler and doesn't require changing the renderer.

Just thinking aloud, but this is a problem I ran into, though it was easily solved in my case since I was using the PR with a customizable projection matrix.

theraot commented:
An oblique projection is a parallel projection.

Here is a picture from Wikipedia:

![image](https://github.com/godotengine/godot/assets/1212509/76fe628b-2d75-420e-a975-8b972a81f907)

It is commonly used in technical drawing. But it has also seen use in video game art, as most 2D RPGs use some kind of oblique projection (where you see the top face and the front face on full size). A notable case is EarthBound that uses a cabinet projection. 

And thus such projection could be useful for anybody trying to mimic the look of those games with a 3D scenes (consider for example the trick used for The Legend of Zelda: A Link Between Worlds). 

---

But it seems you are not doing that. Instead it seems that what you are doing is an oblique *frustum* projection. Please call it what it is. `PROJECTION_OBLIQUE` is wrong.

qaptoR commented:
I've updated the PR to reflect the alternative approach to applying the oblique algorithm.
Now, there is a flag `use_oblique_frustum` which is currently only available for the perspective and orthogonal cameras because as far I am aware the more general algorithm brought to my attention by @lyuma only works for those two projections.

This new approach also reduces the amount of duplicated code that was used for the oblique camera previously.

I've also updated the test project I posted before with a new scene meant to test oblique algorithm on orthogonal cameras.
Some notes about the project in general
- the `perspective_oblique.tscn` portal viewport paths need to be reset when the scene is loaded: `PortalSide[AB]>mesh>material>shader_parameter/TextureAlbedo>viewport_path`
- the orthogonal_oblique.tscn` portal camera frustums seem to be affected in unreliable ways. The boxes/prisms in the scene are placed such that if they are moved further/closer from the position you find them in when the scene loads will result in them being culled by the far plane.
  - on the good side, the near plane does appear to be culling appropriate to the near plane

[89104_min_test.zip](https://github.com/godotengine/godot/files/14543013/89104_min_test.zip)

Ribiveer commented:
Hi there! I'm currently working on a project that needs portals and mirrors, and I'm in sore need of this feature. I see it's been completely written up and working well, which is awesome! Since it's been a few months since this was finished, I'm wondering: are there's any blocks on merging? I'm quite new to contributing to open-source projects, but if there's anything I can dedicate time to getting this feature integrated as soon as possible, I'd love to know what needs to be done.

qaptoR commented:
@Ribiveer I'm not here to say whether anything can be done sooner or later to merge, I'm new to this process as well, despite having made several PRs over the years, they all end up in stasis and I suspect that is either because a) i don't push the godot team enough to remind them of my features, or b) because my PRs are predominantly new features there is an opposing force to them being merged which is that strive for stability in the core engine which new features tend to disrupt leading to slow integration of an already long backlog.

what I would recommend is getting comfortable building from source, and merging different branches/prs into a single working engine that you call your own. this is what I do, and it means that you can have all the features that you want that are stuck in stasis.

obviously I hope that my work gets included in the engine core eventually and I can be counted officially as a contributor, but I also want the current features of the engine to be robust and bug free, so I am find with the godot team focusing their engine there.

I am also grateful that you appreciate my work, and I hope that if you do choose to use it still in the interim that it serves well enough for the purpose you intend.

Ribiveer commented:
@qaptoR Thank you for your thoughtful reply! I think I've been delaying the inevitable, getting comfortable building from source is indeed the best strategy here. In the meantime I hope for you that your PR's get approved! I've looked through your repository and they seem like useful additions indeed.

Ribiveer commented:
I've copied this pull request's commit in [my fork](https://github.com/Ribiveer/godot/tree/ribiveer), and have been testing it today. 

It's working extremely well! It surprised me that the given coordinates were in world space. While this perfectly serves most purposes this feature will used for (most of the time it will be used to align the clipping plane with an object in the world), an argument could also be made that, since all other camera parameters are local to the camera, this one should be as well. Personally though, this is perfect for my (and I assume most people's) needs!

It doesn't seem to play nicely with the automatic LOD system, however. Looking at the clipping plane from straight on, Godot assumes a normal distance for meshes. But when looking at the clipping plane at an angle, Godot seems to assume more and more radical distances, causing objects to assume a lower LOD than they should. I looked around and found through issue #81995 and #76436 that might be related, but further investigation would be necessary for implementation in Godot.

For my own purposes, this PR perfectly suits my needs, since I am not planning on using Godot's built-in LOD system. For integration into the engine there's either a fix needed for working with the LOD system, or it might actually work itself out when the LOD system gets worked out.

Thank you so much for making this PR!

theraot commented:
Now that 4.3 is out, there is another chance for this feature.

It seems everything from prior reviews has been addressed. So I hope it goes smoothly.

Sooshiko commented:
This feature would be very helpful for water reflection as you can go through water like a portal even though your position doesn't change. Especially if it works for compatibility renderer as we don't have access to screen_space reflection.

Riordan-DC commented:
Awesome looking PR! Really looking forward to its merge. Fixes this awful artefact with planar reflections. Pink geometry should be clipped by the planar reflection camera which should be possible with an oblique camera.
![image](https://github.com/user-attachments/assets/3f5522dd-5c3f-4d6f-a4ae-d486b36edd34)

Mickeon commented on doc/classes/Camera3D.xml:
```diff
@@ -202,12 +209,24 @@
 			The camera's size in meters measured as the diameter of the width or height, depending on [member keep_aspect]. Only applicable in orthogonal and frustum modes.
 		</member>
+		<member name="use_oblique_frustum" type="bool" setter="set_use_oblique_frustum" getter="get_use_oblique_frustum" default="false">
+			Toggle for applying oblique near plane frustum culling.
```
```suggestion
			If [code]true[/code], enables oblique near plane frustum culling.
```

qaptoR replied:
change applied for future PR

Mickeon commented on doc/classes/Camera3D.xml:
```diff
@@ -122,6 +122,13 @@
+			<return type="void" />
+			<param index="0" name="oblique_transform" type="Transform3D" />
+			<description>
+				Sets the [member oblique_normal] and [member oblique_position] values of an oblique projection camera using a the origin and forward vector of the transform argument. For ease of using plane meshes as portals and mirrors.
```
```suggestion
				Sets the [member oblique_normal] and [member oblique_position] values of this camera, through the given [param oblique_transform]'s origin and forward vector. Can be useful for plane meshes used as portals or mirrors. See also [member use_oblique_frustum].
```

qaptoR replied:
change applied for future PR

Flarkk commented on doc/classes/Camera3D.xml:
```diff
@@ -202,12 +209,24 @@
 			The distance to the near culling boundary for this camera relative to its local Z axis. Lower values allow the camera to see objects more up close to its origin, at the cost of lower precision across the [i]entire[/i] range. Values lower than the default can lead to increased Z-fighting.
 		</member>
+		<member name="oblique_normal" type="Vector3" setter="set_oblique_normal" getter="get_oblique_normal" default="Vector3(0, 1, 0)">
+			The desired normal vector of the world space plane used by an oblique camera as an oblique near clipping plane.
```
Should the normal vector of the mentioned world space plane be provided in World space or in View space to this function ? Either way this should be clarified.

Flarkk replied:
Gave it further thinking, and in the end I believe the oblique plane's normal and position attributes should be expressed in view space for consistency.
In other words these shouldn't vary when the camera moves around (we're setting camera attributes after all, not scene attributes).
You purposefully provide 
`set_oblique_plane_from_transform()` that allows users to set the oblique plane's attributes in world space already.

Flarkk commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -2626,7 +2650,7 @@ void RendererSceneCull::render_camera(const Ref<RenderSceneBuffers> &p_render_bu
-			camera_data.set_camera(transforms[0], projections[0], false, camera->vaspect, jitter, camera->visible_layers);
+			camera_data.set_camera(transforms[0], projections[0], projections[0], false, camera->vaspect, jitter, camera->visible_layers);
 		} else if (view_count == 2) {
 			camera_data.set_multiview_camera(view_count, transforms, projections, false, camera->vaspect);
```
So the Oblique camera is not supported in XR, right ?
If yes, this should be mentioned somewhere, at least in the documentation

Flarkk commented on servers/rendering/renderer_scene_render.cpp:
```diff
@@ -33,17 +33,18 @@
 	main_transform = p_transform;
-	main_projection = p_projection;
+	main_projection = p_main_projection;
+	shadow_projection = p_shadow_projection;
```
I'd rather call it `culling_projection` as it is what it is used for in your code (except the open question of effects).

Flarkk reviewed (commented):
Generally speaking I find this PR on the right track.

My key concern is that we add world space attributes to the camera. This means the camera would look different depending on its position.

The pattern used instead in such cases in Godot is to set the path to another node (the portal) as an attribute of the dependent node (the camera), and makes the latter retrieve its transform when needed.

I believe this design would simplify the `Camera3D` interface even further as only one `NodePath` field (or equivalent) would be added to `Camera3D`. We could call it `Mirror node` or something similar.

Some attention is needed on effects (SSR, Fog, GI, etc...) : do they still work with this PR ?

Also, you should rename the PR as it's not about a new camera type anymore. *See Edit below though*

---

**Edit** - That could even be a new `RemoteCamera3D` type inheriting `Camera3D`, that not only takes the portal object as a `NodePath` but also the main `Camera3D`. `RemoteCamera3D` would adjust its position automatically relative to both the mirror/portal's transform and the main camera position, similarly to (but not exactly as) how `RemoteTransform3D` operates

Flarkk commented:
Last thought : have you tested Lengyel's trick works even when the oblique plane's normal is not in the camera's XZ plane (i.e. when the portal is not vertical relative to the observer) ?
I've the intuition that the perspective matrix obtained then would not be inversible (it would require to also modify the left and right planes i.e. matrix row 0, but the algorithm only changes row 2).
Maybe I'm wrong, but worth testing if not the case already (I only see vertical portals in your demo)

elvisish commented:
Is this being added to 4.7?

AThousandShips commented on scene/3d/camera_3d.h:
```diff
@@ -137,6 +146,11 @@ class Camera3D : public Node3D {
+	void set_oblique_normal(Vector3 p_oblique_normal);
+	void set_oblique_position(Vector3 p_oblique_position);
+	void set_oblique_offset(real_t p_oblique_offset);
+	void set_oblique_plane_from_transform(Transform3D p_oblique_plane_transform);
```
```suggestion
	void set_oblique_plane_from_transform(const Transform3D &p_oblique_plane_transform);
```

AThousandShips commented on scene/3d/camera_3d.cpp:
```diff
@@ -192,6 +202,16 @@ Projection Camera3D::get_camera_projection() const {
 
+Vector4 Camera3D::_get_oblique_plane() const {
+	Transform3D transform = get_global_transform();
+	int dot = int(oblique_normal.dot(oblique_position - transform.origin) >= 0.0f ? 1.0f : -1.0f);
```
```suggestion
	int dot = (oblique_normal.dot(oblique_position - transform.origin) >= 0.0f) ? 1 : -1;
```

qaptoR commented:
I am closing this PR to resubmit for personal reasons.
Before I resubmit I will address any remaining notes that were made for this PR.

innovate-invent commented:
Thank you for your efforts. I am interested in continuing to follow this features progress. Can you please add a link to this PRs conversation with the new PR so I can subscribe to that?

Thanks!

qaptoR commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -131,6 +140,18 @@ bool RendererSceneCull::is_camera(RID p_camera) const {
+	Camera *camera = camera_owner.get_or_null(p_camera);
+	ERR_FAIL_NULL_V(camera, Vector4());
+
+	int dot = int(camera->oblique_normal.dot(camera->oblique_position - camera->transform.origin) >= 0.0f ? 1.0f : -1.0f);
+	Vector3 cam_space_pos = camera->transform.xform_inv(camera->oblique_position);
+	Vector3 cam_space_normal = camera->transform.basis.xform_inv(camera->oblique_normal) * dot;
+	real_t cam_space_dst = -cam_space_pos.dot(cam_space_normal) + camera->oblique_offset;
```
Tested this: breaks the effect with the following code snippets:

```
int dot = int(camera->oblique_normal.dot(camera->oblique_position - camera->transform.origin) >= 0.0f ? 1.0f : -1.0f);
Vector3 cam_space_normal = camera->transform.basis.xform_inv(camera->oblique_normal) * dot;
real_t cam_space_dst = -(camera->oblique_position - camera->transform.origin).dot(camera->oblique_normal) + camera->oblique_offset;
```
and
```
int dot = int(camera->oblique_normal.dot(camera->oblique_position - camera->transform.origin) >= 0.0f ? 1.0f : -1.0f);
Vector3 cam_space_normal = camera->oblique_normal * dot;
real_t cam_space_dst = -(camera->oblique_position - camera->transform.origin).dot(camera->oblique_normal) + camera->oblique_offset;
```
for the record I just left the name 'cam_space_normal' for clarity while testing, even if its not quite that same vector in snippet 2

both versions look pretty much like this, with slight variation:
<img width="1682" height="969" alt="image" src="https://github.com/user-attachments/assets/d674d927-640c-4a3a-9626-877e0baef053" />

for the record, it should look like this:
<img width="1682" height="969" alt="image" src="https://github.com/user-attachments/assets/12be2b44-cf22-4367-8625-903bd2dd6af6" />

qaptoR commented:
re: https://github.com/godotengine/godot/pull/89140#discussion_r1870270884

Definitely breaks right now for Bokeh and GI (only tested so far). GI specifically breaks on Projection::invert() for some reason that I'm looking into:

```error message
ERROR: Condition "0.0 == r2[2]" is true.
   at: invert (core/math/projection.cpp:718)
```

I'm going to try my best to get all these test scenarios working before I make another PR, but I'm not going to give it more than a couple of weeks of effort, and then I might need to ask for expertise on how rendering works. Shadows was easy-ish to fix on my own previously because the relevant shadow rendering code was reasonably named when grepping the source files. But for the other effects, I'll have to really hunt down where the projections effect their rendering.
that said, if anyone reading this knows off the top of their head (don't go looking otherwise) where those places are, I'd appreciate the direction. I would ask for more help (and its genuinely not an ego thing), I just don't want to bother anyone with what is essentially a special interest of mine.

For those interested, I've ported this PR into a clean repo [here](https://github.com/qaptoR-godot/godot/tree/master_oblique_frustum_camera). Feel free to take a look and make a PR if you solve some of the existing effects issues before me. For reference, I'm using the same test scenes that were provided before, but here they are again for reference: 
[reprojection.zip](https://github.com/user-attachments/files/29751923/reprojection.zip)
